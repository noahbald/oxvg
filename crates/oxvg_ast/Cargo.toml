[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
