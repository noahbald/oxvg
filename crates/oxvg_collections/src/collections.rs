use std::collections::{BTreeMap, BTreeSet};

pub trait Group<'a> {
    fn matches<T>(&self, value: T) -> bool
    where
        T: Into<&'a str>,
    {
        self.set().contains(value.into())
    }

    fn set(&self) -> &'a BTreeSet<&'static str>;
}

#[derive(Hash, PartialEq, Eq)]
pub enum ElementGroup {
    Animation,
    Descriptive,
    Shape,
    Structural,
    PaintServer,
    NonRendering,
    Container,
    TextContent,
    TextContentChild,
    LightSource,
    FilterPrimitive,
}

impl<'a> Group<'a> for ElementGroup {
    fn set(&self) -> &'a BTreeSet<&'static str> {
        match self {
            Self::Animation => &ANIMATION,
            Self::Descriptive => &DESCRIPTIVE,
            Self::Shape => &SHAPE,
            Self::Structural => &STRUCTURAL,
            Self::PaintServer => &PAINT_SERVER,
            Self::NonRendering => &NON_RENDERING,
            Self::Container => &CONTAINER,
            Self::TextContent => &TEXT_CONTENT,
            Self::TextContentChild => &TEXT_CONTENT_CHILD,
            Self::LightSource => &LIGHT_SOURCE,
            Self::FilterPrimitive => &FILTER_PRIMITIVE,
        }
    }
}

// Element groups
lazy_static! {
    static ref ANIMATION: BTreeSet<&'static str> = BTreeSet::from([
        "animate",
        "animateColor",
        "animateMotion",
        "animateTransform",
        "set"
    ]);
    static ref DESCRIPTIVE: BTreeSet<&'static str> = BTreeSet::from(["desc", "metadata", "title"]);
    static ref SHAPE: BTreeSet<&'static str> =
        BTreeSet::from(["circle", "ellipse", "line", "path", "polygon", "polyline", "rect"]);
    static ref STRUCTURAL: BTreeSet<&'static str> =
        BTreeSet::from(["defs", "g", "svg", "symbol", "use"]);
    static ref PAINT_SERVER: BTreeSet<&'static str> = BTreeSet::from([
        "hatch",
        "linearGradient",
        "meshGradient",
        "pattern",
        "radialGradient",
        "solidColor"
    ]);
    static ref NON_RENDERING: BTreeSet<&'static str> = BTreeSet::from([
        "clipPath",
        "filter",
        "linearGradient",
        "marker",
        "mask",
        "pattern",
        "radialGradient",
        "solidColor",
        "symbol"
    ]);
    static ref CONTAINER: BTreeSet<&'static str> = BTreeSet::from([
        "a",
        "defs",
        "foreignObject",
        "g",
        "marker",
        "mask",
        "missing-glyph",
        "pattern",
        "svg",
        "switch",
        "symbol"
    ]);
    static ref TEXT_CONTENT: BTreeSet<&'static str> = BTreeSet::from([
        "a",
        "altGlyph",
        "altGlyphDef",
        "alyGlyphItem",
        "glyph",
        "glyphRef",
        "text",
        "textPath",
        "tref",
        "tspan"
    ]);
    static ref TEXT_CONTENT_CHILD: BTreeSet<&'static str> =
        BTreeSet::from(["alyGlyph", "textPath", "tref", "tspan"]);
    static ref LIGHT_SOURCE: BTreeSet<&'static str> = BTreeSet::from([
        "feDiffuseLighting",
        "feDistantLight",
        "fePointLight",
        "feSpecularLighting",
        "feSpotLight"
    ]);
    static ref FILTER_PRIMITIVE: BTreeSet<&'static str> = BTreeSet::from([
        "feBlend",
        "feColorMatrix",
        "feComponentTransfer",
        "feComposite",
        "feConvolveMatrix",
        "feDiffuseLighting",
        "feDisplacementMap",
        "feDropShadow",
        "feFlood",
        "feFuncA",
        "feFuncB",
        "feFuncG",
        "feFuncR",
        "feGaussianBlur",
        "feImage",
        "feMerge",
        "feMergeNode",
        "feMorphology",
        "feOffset",
        "feSpecularLighting",
        "feTile",
        "feTurbulence"
    ]);
}

#[derive(Hash, PartialEq, Eq)]
pub enum AttrsGroups {
    AnimationAddition,
    AnimationAttributeTarget,
    AnimationEvent,
    AnimationTiming,
    AnimationValue,
    ConditionalProcessing,
    Core,
    GraphicalEvent,
    Presentation,
    XLink,
    DocumentEvent,
    DocumentElementEvent,
    GlobalEvent,
    FilterPrimitive,
    TransferFunction,
}

impl<'a> Group<'a> for AttrsGroups {
    fn set(&self) -> &'a BTreeSet<&'static str> {
        match self {
            Self::AnimationAddition => &ANIMATION_ADDITION,
            Self::AnimationAttributeTarget => &ANIMATION_ATTRIBUTE_TARGET,
            Self::AnimationEvent => &ANIMATION_EVENT,
            Self::AnimationTiming => &ANIMATION_TIMING,
            Self::AnimationValue => &ANIMATION_VALUE,
            Self::ConditionalProcessing => &CONDITIONAL_PROCESSING,
            Self::Core => &CORE,
            Self::GraphicalEvent => &GRAPHICAL_EVENT,
            Self::Presentation => &PRESENTATION,
            Self::XLink => &X_LINK,
            Self::DocumentEvent => &DOCUMENT_EVENT,
            Self::DocumentElementEvent => &DOCUMENT_EVEMENT_EVENT,
            Self::GlobalEvent => &GLOBAL_EVENT,
            Self::FilterPrimitive => &FILTER_PRIMITIVE,
            Self::TransferFunction => &TRANSFER_FUNCTION,
        }
    }
}

// Attrs groups
lazy_static! {
    pub static ref ANIMATION_ADDITION: BTreeSet<&'static str> =
        BTreeSet::from(["additive", "accumulate"]);
    pub static ref ANIMATION_ATTRIBUTE_TARGET: BTreeSet<&'static str> =
        BTreeSet::from(["attributeType", "attributeName"]);
    pub static ref ANIMATION_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["onbegin", "onend", "onrepeat", "onload"]);
    pub static ref ANIMATION_TIMING: BTreeSet<&'static str> = BTreeSet::from([
        "begin",
        "dur",
        "end",
        "fill",
        "max",
        "min",
        "repeatCount",
        "repeatDur",
        "restart",
    ]);
    pub static ref ANIMATION_VALUE: BTreeSet<&'static str> = BTreeSet::from([
        "by",
        "calcMode",
        "from",
        "keySplines",
        "keyTimes",
        "to",
        "values",
    ]);
    pub static ref CONDITIONAL_PROCESSING: BTreeSet<&'static str> =
        BTreeSet::from(["requiredExtensions", "requiredFeatures", "systemLanguage",]);
    pub static ref CORE: BTreeSet<&'static str> =
        BTreeSet::from(["id", "tabindex", "xml:base", "xml:lang", "xml:space"]);
    pub static ref GRAPHICAL_EVENT: BTreeSet<&'static str> = BTreeSet::from([
        "onactivate",
        "onclick",
        "onfocusin",
        "onfocusout",
        "onload",
        "onmousedown",
        "onmousemove",
        "onmouseout",
        "onmouseover",
        "onmouseup",
    ]);
    pub static ref PRESENTATION: BTreeSet<&'static str> = BTreeSet::from([
        "alignment-baseline",
        "baseline-shift",
        "clip-path",
        "clip-rule",
        "clip",
        "color-interpolation-filters",
        "color-interpolation",
        "color-profile",
        "color-rendering",
        "color",
        "cursor",
        "direction",
        "display",
        "dominant-baseline",
        "enable-background",
        "fill-opacity",
        "fill-rule",
        "fill",
        "filter",
        "flood-color",
        "flood-opacity",
        "font-family",
        "font-size-adjust",
        "font-size",
        "font-stretch",
        "font-style",
        "font-variant",
        "font-weight",
        "glyph-orientation-horizontal",
        "glyph-orientation-vertical",
        "image-rendering",
        "letter-spacing",
        "lighting-color",
        "marker-end",
        "marker-mid",
        "marker-start",
        "mask",
        "opacity",
        "overflow",
        "paint-order",
        "pointer-events",
        "shape-rendering",
        "stop-color",
        "stop-opacity",
        "stroke-dasharray",
        "stroke-dashoffset",
        "stroke-linecap",
        "stroke-linejoin",
        "stroke-miterlimit",
        "stroke-opacity",
        "stroke-width",
        "stroke",
        "text-anchor",
        "text-decoration",
        "text-overflow",
        "text-rendering",
        "transform-origin",
        "transform",
        "unicode-bidi",
        "vector-effect",
        "visibility",
        "word-spacing",
        "writing-mode",
    ]);
    pub static ref X_LINK: BTreeSet<&'static str> = BTreeSet::from([
        "xlink:actuate",
        "xlink:arcrole",
        "xlink:href",
        "xlink:role",
        "xlink:show",
        "xlink:title",
        "xlink:type",
    ]);
    pub static ref DOCUMENT_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["onabort", "onerror", "onresize", "onscroll", "onunload", "onzoom",]);
    pub static ref DOCUMENT_EVEMENT_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["oncopy", "oncut", "onpaste"]);
    pub static ref GLOBAL_EVENT: BTreeSet<&'static str> = BTreeSet::from([
        "oncancel",
        "oncanplay",
        "oncanplaythrough",
        "onchange",
        "onclick",
        "onclose",
        "oncuechange",
        "ondblclick",
        "ondrag",
        "ondragend",
        "ondragenter",
        "ondragleave",
        "ondragover",
        "ondragstart",
        "ondrop",
        "ondurationchange",
        "onemptied",
        "onended",
        "onerror",
        "onfocus",
        "oninput",
        "oninvalid",
        "onkeydown",
        "onkeypress",
        "onkeyup",
        "onload",
        "onloadeddata",
        "onloadedmetadata",
        "onloadstart",
        "onmousedown",
        "onmouseenter",
        "onmouseleave",
        "onmousemove",
        "onmouseout",
        "onmouseover",
        "onmouseup",
        "onmousewheel",
        "onpause",
        "onplay",
        "onplaying",
        "onprogress",
        "onratechange",
        "onreset",
        "onresize",
        "onscroll",
        "onseeked",
        "onseeking",
        "onselect",
        "onshow",
        "onstalled",
        "onsubmit",
        "onsuspend",
        "ontimeupdate",
        "ontoggle",
        "onvolumechange",
        "onwaiting",
    ]);
    pub static ref FILTER_PRIMITIVE_ATTRS: BTreeSet<&'static str> =
        BTreeSet::from(["x", "y", "width", "height", "result"]);
    pub static ref TRANSFER_FUNCTION: BTreeSet<&'static str> = BTreeSet::from([
        "amplitude",
        "exponent",
        "intercept",
        "offset",
        "slope",
        "tableValues",
        "type",
    ]);
}

impl AttrsGroups {
    pub fn defaults(&self) -> Option<&BTreeMap<&'static str, &'static str>> {
        match self {
            Self::Core => Some(&CORE_DEFAULTS),
            Self::Presentation => Some(&PRESENTATION_DEFAULTS),
            Self::TransferFunction => Some(&TRANSFER_FUNCTION_DEFAULTS),
            _ => None,
        }
    }
}

// Attrs groups defaults
lazy_static! {
    static ref CORE_DEFAULTS: BTreeMap<&'static str, &'static str> =
        BTreeMap::from([("xml:space", "default")]);
    static ref PRESENTATION_DEFAULTS: BTreeMap<&'static str, &'static str> = BTreeMap::from([
        ("clip", "auto"),
        ("clip-path'", "none"),
        ("clip-rule'", "nonzero"),
        ("mask", "none"),
        ("opacity", "1"),
        ("stop-color'", "#000"),
        ("stop-opacity'", "1"),
        ("fill-opacity'", "1"),
        ("fill-rule'", "nonzero"),
        ("fill", "#000"),
        ("stroke", "none"),
        ("stroke-width'", "1"),
        ("stroke-linecap'", "butt"),
        ("stroke-linejoin'", "miter"),
        ("stroke-miterlimit'", "4"),
        ("stroke-dasharray'", "none"),
        ("stroke-dashoffset'", "0"),
        ("stroke-opacity'", "1"),
        ("paint-order'", "normal"),
        ("vector-effect'", "none"),
        ("display", "inline"),
        ("visibility", "visible"),
        ("marker-start'", "none"),
        ("marker-mid'", "none"),
        ("marker-end'", "none"),
        ("color-interpolation'", "sRGB"),
        ("color-interpolation-filters'", "linearRGB"),
        ("color-rendering'", "auto"),
        ("shape-rendering'", "auto"),
        ("text-rendering'", "auto"),
        ("image-rendering'", "auto"),
        ("font-style'", "normal"),
        ("font-variant'", "normal"),
        ("font-weight'", "normal"),
        ("font-stretch'", "normal"),
        ("font-size'", "medium"),
        ("font-size-adjust'", "none"),
        ("kerning", "auto"),
        ("letter-spacing'", "normal"),
        ("word-spacing'", "normal"),
        ("text-decoration'", "none"),
        ("text-anchor'", "start"),
        ("text-overflow'", "clip"),
        ("writing-mode'", "lr-tb"),
        ("glyph-orientation-vertical'", "auto"),
        ("glyph-orientation-horizontal'", "0deg"),
        ("direction", "ltr"),
        ("unicode-bidi'", "normal"),
        ("dominant-baseline'", "auto"),
        ("alignment-baseline'", "baseline"),
        ("baseline-shift'", "baseline"),
    ]);
    static ref TRANSFER_FUNCTION_DEFAULTS: BTreeMap<&'static str, &'static str> = BTreeMap::from([
        ("slope", "1"),
        ("intercept", "0"),
        ("amplitude", "1"),
        ("exponent", "1"),
        ("offset", "0"),
    ]);
}

lazy_static! {
    pub static ref PATH_ELEMS: BTreeSet<&'static str> =
        BTreeSet::from(["glyph", "missing-glyph", "path"]);
    pub static ref INHERITABLE_ATTRS: BTreeSet<&'static str> = BTreeSet::from([
        "clip-rule",
        "color-interpolation-filters",
        "color-interpolation",
        "color-profile",
        "color-rendering",
        "color",
        "cursor",
        "direction",
        "dominant-baseline",
        "fill-opacity",
        "fill-rule",
        "fill",
        "font-family",
        "font-size-adjust",
        "font-size",
        "font-stretch",
        "font-style",
        "font-variant",
        "font-weight",
        "font",
        "glyph-orientation-horizontal",
        "glyph-orientation-vertical",
        "image-rendering",
        "letter-spacing",
        "marker-end",
        "marker-mid",
        "marker-start",
        "marker",
        "paint-order",
        "pointer-events",
        "shape-rendering",
        "stroke-dasharray",
        "stroke-dashoffset",
        "stroke-linecap",
        "stroke-linejoin",
        "stroke-miterlimit",
        "stroke-opacity",
        "stroke-width",
        "stroke",
        "text-anchor",
        "text-rendering",
        "transform",
        "visibility",
        "word-spacing",
        "writing-mode",
    ]);
    pub static ref PRESENTATION_NON_INHERITABLE_GROUP_ATTRS: BTreeSet<&'static str> =
        BTreeSet::from([
            "clip-path",
            "display",
            "filter",
            "mask",
            "opacity",
            "text-decoration",
            "transform",
            "unicode-bidi",
        ]);
    pub static ref REFERENCES_PROPS: BTreeSet<&'static str> = BTreeSet::from([
        "clip-path",
        "color-profile",
        "fill",
        "filter",
        "marker-end",
        "marker-mid",
        "marker-start",
        "mask",
        "stroke",
        "style",
    ]);
    pub static ref PSEUDO_DISPLAY_STATE: BTreeSet<&'static str> =
        BTreeSet::from(["fullscreen", "modal", "picture-in-picture"]);
    pub static ref PSEUDO_INPUT: BTreeSet<&'static str> = BTreeSet::from([
        "autofill",
        "blank",
        "checked",
        "default",
        "disabled",
        "enabled",
        "in-range",
        "indetermined",
        "invalid",
        "optional",
        "out-of-range",
        "placeholder-shown",
        "read-only",
        "read-write",
        "required",
        "user-invalid",
        "valid",
    ]);
    pub static ref PSEUDO_LINGUISTIC: BTreeSet<&'static str> = BTreeSet::from(["dir", "lang"]);
    pub static ref PSEUDO_LOCATION: BTreeSet<&'static str> = BTreeSet::from([
        "any-link",
        "link",
        "local-link",
        "scope",
        "target-within",
        "target",
        "visited",
    ]);
    pub static ref PSEUDO_RESOURCE_STATE: BTreeSet<&'static str> =
        BTreeSet::from(["playing", "paused"]);
    pub static ref PSEUDO_TIME_DIMENSIONAL: BTreeSet<&'static str> =
        BTreeSet::from(["current", "past", "future"]);
    pub static ref PSEUDO_TREE_STRUCTURAL: BTreeSet<&'static str> = BTreeSet::from([
        "empty",
        "first-child",
        "first-of-type",
        "last-child",
        "last-of-type",
        "nth-child",
        "nth-last-child",
        "nth-last-of-type",
        "nth-of-type",
        "only-child",
        "only-of-type",
        "root",
    ]);
    pub static ref PSEUDO_USER_ACTION: BTreeSet<&'static str> =
        BTreeSet::from(["active", "focus-visible", "focus-within", "focus", "hover",]);
    pub static ref PSEUDO_FUNCTIONAL: BTreeSet<&'static str> =
        BTreeSet::from(["is", "not", "where", "has"]);
}
