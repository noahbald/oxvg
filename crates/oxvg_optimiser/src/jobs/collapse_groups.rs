use lightningcss::{
    properties::PropertyId,
    stylesheet::{ParserOptions, StyleAttribute},
    vendor_prefix::VendorPrefix,
};
use oxvg_ast::{
    attribute::{Attr, Attributes},
    element::Element,
    name::Name,
    node::Node,
};
use oxvg_derive::OptionalDefault;
use oxvg_selectors::collections::{ElementGroup, Group, INHERITABLE_ATTRS};
use serde::Deserialize;

use crate::{Context, Job, JobDefault, PrepareOutcome};

#[derive(Deserialize, Clone, OptionalDefault)]
#[serde(rename_all = "camelCase")]
pub struct CollapseGroups(bool);

impl Job for CollapseGroups {
    fn prepare(&mut self, _document: &impl Node) -> crate::PrepareOutcome {
        if self.0 {
            PrepareOutcome::None
        } else {
            PrepareOutcome::Skip
        }
    }

    fn run(&self, element: &impl Element, _context: &Context) {
        let Some(parent) = Element::parent_element(element) else {
            return;
        };

        if element.is_root() || parent.local_name().as_ref() == "switch" {
            return;
        }
        if element.local_name().as_ref() != "g" || !element.has_child_elements() {
            return;
        }

        move_attributes_to_child(element);
        flatten_when_all_attributes_moved(element);
    }
}

impl Default for CollapseGroups {
    fn default() -> Self {
        Self(true)
    }
}

fn move_attributes_to_child(element: &impl Element) {
    log::debug!("collapse_groups: move_attributes_to_child");

    let mut children = element.children_iter();
    let Some(first_child) = children.next() else {
        log::debug!("collapse_groups: not moving attrs: no children");
        return;
    };
    if children.next().is_some() {
        log::debug!("collapse_groups: not moving attrs: many children");
        return;
    }

    let attrs = element.attributes();
    if attrs.len() == 0 {
        log::debug!("collapse_groups: not moving attrs: no attrs to move");
        return;
    }

    if dbg!(is_group_identifiable(element, &first_child))
        || dbg!(is_position_visually_unstable(element, &first_child))
        || dbg!(is_node_with_filter(element))
    {
        log::debug!("collapse_groups: not moving attrs: see true condition");
        return;
    };

    let mut removals = Vec::default();
    let first_child_attrs = first_child.attributes();
    for attr in attrs.iter() {
        let name = attr.name();
        let local_name = name.local_name();
        let local_name: &str = local_name.as_ref();
        let value = attr.value();

        let child_attr = first_child_attrs.get_named_item(&name);
        if has_animated_attr(&first_child, &name) {
            log::debug!("collapse_groups: canelled moves: has animated_attr");
            return;
        }

        removals.push(attr.name());
        let Some(mut child_attr) = child_attr else {
            log::debug!("collapse_groups: moved {name}: same as parent",);
            first_child_attrs.set_named_item(attr.into_owned());
            continue;
        };

        let child_attr_value = child_attr.value();
        if name.local_name().as_ref() == "transform" {
            log::debug!("collapse_groups: moved transform: is transform");
            child_attr.set_value(format!("{value} {child_attr_value}").into());
        } else if child_attr_value.as_ref() == "inherit" {
            log::debug!("collapse_groups: moved {name}: is explicit inherit");
            child_attr.set_value(value);
        } else if !INHERITABLE_ATTRS.contains(local_name) && child_attr.value() != value {
            log::debug!("collapse_groups: removing {name}: inheritable attr is not inherited");
            return;
        }
    }

    for attr in removals {
        element.remove_attribute(&attr);
    }
}

fn flatten_when_all_attributes_moved(element: &impl Element) {
    if !element.attributes().is_empty() {
        log::debug!("skipping flatten: has attributes");
        return;
    }

    let animation_group = oxvg_selectors::collections::Group::set(&ElementGroup::Animation);
    {
        if element.depth_first().any(|child| {
            let local_name = child.local_name();
            let name: &str = local_name.as_ref();
            dbg!(animation_group.contains(dbg!(name)))
        }) {
            log::debug!("skipping flatten: has animating child");
            return;
        }
    }

    element.flatten();
}

fn has_animated_attr(element: &impl Element, name: &impl Name) -> bool {
    let local_name = name.local_name();
    let node_name: &str = local_name.as_ref();
    for child in std::iter::once(element.clone()).chain(element.depth_first()) {
        let child_name_local = child.local_name();
        let child_name: &str = child_name_local.as_ref();
        if Group::set(&ElementGroup::Animation).contains(child_name)
            && child
                .get_attribute_local(&"attributeName".into())
                .is_some_and(|attr| attr.as_ref() == node_name)
        {
            return true;
        }
    }
    false
}

fn is_group_identifiable<E: Element>(node: &E, child: &E) -> bool {
    let class = &"class".into();
    child.has_attribute_local(&"id".into())
        && (!node.has_attribute_local(class) || !child.has_attribute_local(class))
}

fn is_position_visually_unstable<E: Element>(node: &E, child: &E) -> bool {
    let is_node_clipping =
        node.has_attribute_local(&"clip-path".into()) || node.has_attribute_local(&"mask".into());
    let is_child_transformed_group =
        child.local_name().as_ref() == "g" && child.has_attribute_local(&"transform".into());
    is_node_clipping || is_child_transformed_group
}

fn is_node_with_filter(node: &impl Element) -> bool {
    node.has_attribute_local(&"filter".into())
        || node
            .get_attribute_local(&"style".into())
            .is_some_and(|code| {
                StyleAttribute::parse(code.as_ref(), ParserOptions::default()).is_ok_and(|style| {
                    style
                        .declarations
                        .get(&PropertyId::Filter(VendorPrefix::None))
                        .is_some()
                })
            })
}

#[test]
#[allow(clippy::too_many_lines)]
fn collapse_groups() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove both useless `g`s -->
    <g>
        <g>
            <path d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should pass all inheritable attributes to children -->
    <g>
        <g attr1="val1">
            <path d="..."/>
        </g>
    </g>
    <g attr1="val1">
        <g attr2="val2">
            <path d="..."/>
        </g>
    </g>
    <g attr1="val1">
        <g>
            <path d="..."/>
        </g>
        <path d="..."/>
    </g>
    <g attr1="val1">
        <g attr2="val2">
            <path d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove inheritable overridden attributes -->
    <g attr1="val1">
        <g fill="red">
            <path fill="green" d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove group with equal attribute values to child -->
    <g attr1="val1">
        <g attr2="val2">
            <path attr2="val2" d="..."/>
        </g>
        <g attr2="val2">
            <path attr2="val3" d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should join transform attributes into `transform="rotate(45) scale(2)"` -->
    <g attr1="val1">
        <g transform="rotate(45)">
            <path transform="scale(2)" d="..."/>
        </g>
        <path d="..."/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `clip-path` -->
    <clipPath id="a">
       <path d="..."/>
    </clipPath>
    <clipPath id="b">
       <path d="..."/>
    </clipPath>
    <g transform="matrix(0 -1.25 -1.25 0 100 100)" clip-path="url(#a)">
        <g transform="scale(.2)">
            <path d="..."/>
            <path d="..."/>
        </g>
    </g>
    <g transform="matrix(0 -1.25 -1.25 0 100 100)" clip-path="url(#a)">
        <g transform="scale(.2)">
            <g>
                <g clip-path="url(#b)">
                    <path d="..."/>
                    <path d="..."/>
                </g>
            </g>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `clip-path` and `mask` -->
    <clipPath id="a">
       <path d="..."/>
    </clipPath>
    <path d="..."/>
    <g clip-path="url(#a)">
        <path d="..." transform="scale(.2)"/>
    </g>
    <g mask="url(#a)">
        <path d="..." transform="scale(.2)"/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with `id` or animation children -->
    <g stroke="#000">
        <g id="star">
            <path id="bar" d="..."/>
        </g>
    </g>
    <g>
        <animate id="frame0" attributeName="visibility" values="visible" dur="33ms" begin="0s;frame27.end"/>
        <path d="..." fill="#272727"/>
        <path d="..." fill="#404040"/>
        <path d="..." fill="#2d2d2d"/>
    </g>
    <g transform="rotate(-90 25 0)">
        <circle stroke-dasharray="110" r="20" stroke="#10cfbd" fill="none" stroke-width="3" stroke-linecap="round">
            <animate attributeName="stroke-dashoffset" values="360;140" dur="2.2s" keyTimes="0;1" calcMode="spline" fill="freeze" keySplines="0.41,0.314,0.8,0.54" repeatCount="indefinite" begin="0"/>
            <animateTransform attributeName="transform" type="rotate" values="0;274;360" keyTimes="0;0.74;1" calcMode="linear" dur="2.2s" repeatCount="indefinite" begin="0"/>
        </circle>
    </g>
</svg>"##
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve groups with classes -->
    <style>
        .n{display:none}
        .i{display:inline}
    </style>
    <g id="a">
        <g class="i"/>
    </g>
    <g id="b" class="n">
        <g class="i"/>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should preserve children of `<switch>` -->
    <switch>
        <g id="a">
            <g class="i"/>
        </g>
        <g id="b" class="n">
            <g class="i"/>
        </g>
        <g>
            <g/>
        </g>
    </switch>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should replace inheritable value -->
	<g color="red">
		<g color="inherit" fill="none" stroke="none">
			<circle cx="130" cy="80" r="60" fill="currentColor"/>
			<circle cx="350" cy="80" r="60" stroke="currentColor" stroke-width="4"/>
		</g>
	</g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove useless group -->
    <g filter="url(#...)">
        <g>
            <path d="..."/>
        </g>
    </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 88 88">
  <!-- Should preserve group if some attrs cannot be moved -->
  <filter id="a">
    <feGaussianBlur stdDeviation="1"/>
  </filter>
  <g transform="matrix(0.6875,0,0,0.6875,20.34375,66.34375)" style="filter:url(#a)">
    <path d="M 33.346591,-83.471591 L -10.744318,-36.471591 L -10.49989,-32.5" style="fill-opacity:1"/>
  </g>
</svg>"#
        )
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "collapseGroups": true }"#,
        Some(
            r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100">
    <!-- Should preserve group if parent has `filter` -->
    <clipPath id="a">
        <circle cx="25" cy="15" r="10"/>
    </clipPath>
    <filter id="b">
        <feColorMatrix type="saturate"/>
    </filter>
    <g filter="url(#b)">
        <g clip-path="url(#a)">
            <circle cx="30" cy="10" r="10" fill="yellow" id="c1"/>
        </g>
    </g>
    <g style="filter:url(#b)">
        <g clip-path="url(#a)">
            <circle cx="20" cy="10" r="10" fill="blue" id="c2"/>
        </g>
    </g>
    <circle cx="25" cy="15" r="10" stroke="black" stroke-width=".1" fill="none"/>
</svg>"#
        )
    )?);

    Ok(())
}
