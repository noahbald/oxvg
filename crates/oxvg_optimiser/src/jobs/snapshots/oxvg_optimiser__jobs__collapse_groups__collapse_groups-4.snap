---
source: crates/oxvg_optimiser/src/jobs/collapse_groups.rs
expression: "test_config(r#\"{ \"collapseGroups\": true }\"#,\nSome(r#\"<svg xmlns=\"http://www.w3.org/2000/svg\">\n    <!-- Should remove group with equal attribute values to child -->\n    <g attr1=\"val1\">\n        <g attr2=\"val2\">\n            <path attr2=\"val2\" d=\"...\"/>\n        </g>\n        <g attr2=\"val2\">\n            <path attr2=\"val3\" d=\"...\"/>\n        </g>\n    </g>\n</svg>\"#))?"
---
<svg xmlns="http://www.w3.org/2000/svg">
    <!-- Should remove group with equal attribute values to child -->
    <g attr1="val1">
        
            <path attr2="val2" d="..."></path>
        
        <g attr2="val2">
            <path attr2="val3" d="..."></path>
        </g>
    </g>
</svg>
